### PR TITLE
ATO-130: Refactor validation of requests to /authorize

### PR DESCRIPTION
**Pull request by billrdunn**
## What?

- Extract query params validation logic out of `OrchestrationAuthorizationService` into new class `QueryParamsValidationService`. 
- Rename `RequestObjectService` to `RequestObjectValidationService`.
- Create `AbstractValidationService` which `QueryParamsValidationService` and `RequestObjectValidationService` extend.

## Why?

- Didn't make sense for `OrchestrationAuthorizationService` to handle query parameter validation, while JAR validation was handled by `RequestObjectService`.
- `OrchestrationAuthorizationService` (and corresponding test file) was doing too many things.

### CODE REVIEW

**Comment by mrwilson** [comment]
Looks good! We don't necessarily have to change it in this PR, but I think that we could remove the dependency on `IpvCapacityService` from both validators to a separate component / method.

My thinking is that it is perfectly possible to have a valid authorisation request, whether query param or request object, and for IPV to just be switched off.

**Review by kerrr — APPROVED** [review]
Very happy to just roll forward, but some thoughts:

- Great to see the rename refactors in their own commits, makes it much easier to see the real change.
- We seem to have loads of things named `*Service` and I'm not sure how many are "services" (whatever that is) to my way of thinking. Would `QueryParamsValidationService` be better thought of as just a `QueryParamsValidator`?
- I'm always cautious about using inheritance. It has it's place, but in many cases is not the best factoring. Does `AbstractValidationService` make sense here? Or is it just a way of deduping a bit of boiler plate in the constructors and bit of shared client registry code?  Would another factoring be clearer?
- I suspect that some deeper refactoring would clarify and simplify this a _lot_ and make some of this obsolete, but this is a good step in that direction.

🚀

**Comment by mrwilson** [comment]
+1 to Kerr's comment about `*Service` naming, I've raised this with Ethan before: this is a hangover from the auth-only MVP where the main service classes were wrappers around AWS components like KMS, Dynamo, etc.

It's not a _huge_ thing, but I am in favour of making names clearer where we can.

**Comment by billrdunn** [comment]
> Looks good! We don't necessarily have to change it in this PR, but I think that we could remove the dependency on `IpvCapacityService` from both validators to a separate component / method.
> 
> My thinking is that it is perfectly possible to have a valid authorisation request, whether query param or request object, and for IPV to just be switched off.

@mrwilson  I've made a separate ticket (ATO-169)

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/AbstractValidator.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import uk.gov.di.authentication.oidc.entity.AuthRequestError;
+import uk.gov.di.authentication.oidc.services.IPVCapacityService;
+import uk.gov.di.authentication.shared.entity.ClientRegistry;
+import uk.gov.di.authentication.shared.services.ConfigurationService;
+import uk.gov.di.authentication.shared.services.DynamoClientService;
+
+import java.util.Objects;
+import java.util.Optional;
+
+public abstract class AbstractValidator {
```
[the remark this answers is not in the document]
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/AbstractValidator.java]: Done

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/WellknownHandler.java [hunk 2]
```diff
@@ -79,6 +79,7 @@ private String constructProviderMetadata(ConfigurationService configService) {
             oidcMetadata.setClaimTypes(List.of(ClaimType.NORMAL));
             oidcMetadata.setClaims(ValidClaims.allOneLoginClaims());
             oidcMetadata.setSupportsRequestURIParam(false);
+            oidcMetadata.setSupportsRequestParam(true);
```
[the remark this answers is not in the document]
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/WellknownHandler.java]: Sorted

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsValidator.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        if (authRequest.getRequestURI() != null) {
+            LOG.error("Request URI is not supported");
+            return Optional.of(
+                    new AuthRequestError(OAuth2Error.REQUEST_URI_NOT_SUPPORTED, redirectURI));
+        }
+        if (authRequest.getRequestObject() != null) {
+            LOG.error("Request object not expected here");
+            return Optional.of(
+                    new AuthRequestError(OAuth2Error.REQUEST_NOT_SUPPORTED, redirectURI));
+        }
```
[the remark this answers is not in the document]
Comment (billrdunn) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/validators/QueryParamsValidator.java]: Have removed it

**Review by ethanmills — APPROVED** [review]

**Pull request merged**